### PR TITLE
TraceLog streaming/in-memory EventPipe support

### PR DESCRIPTION
**Pull request by vaind**
I'm trying to make changes to enable `TraceLog` use without creating an ETLX file from an EventPipe (nettrace). I've taken the existing ETW streaming support as a baseline and tried to adapt that. However, I'm having trouble getting the newly added tests to pass because the events coming from TraceLog don't match what I'd get from `EventPipeEventSource` (the previous test case, `Streaming` in the same file). 

@brianrob maybe if you could have a look at my changes, you could spot an issue what I'm doing wrong (or not doing and I should be) to get this working properly? Or is there some bigger problem I'm missing completely that prevents this from happening? 

The use-case I'm going after is a continuously running sampling profiler that only samples as needed (because actually starting the profiler takes a lot of time in the CLR so I want to do it only once).

This is a follow up, though not strictly an implementation of, #1829

### CODE REVIEW

**Comment by vaind** [comment]
@microsoft-github-policy-service agree

**Comment by brianrob** [comment]
@vaind, thanks for working on this.  I think at a high level this should work.  A couple of thoughts:

1. It would be worthwhile to architect this similiar to how we do with `TraceEventSession` for live sessions.  This allows the `TraceLog` to know that it is being used in a live mode instead of a file-based mode.  There are some differences in how one can use the `TraceLog` when its live, because it hasn't seen the "end" of the trace, unlike in the file-based mode.
2. Can you share what differences you're seeing?  It's possible that it's just some piece of data that hasn't made it over to the `TraceLog`.

**Comment by vaind** [comment]
> @vaind, thanks for working on this. I think at a high level this should work. A couple of thoughts:

Thanks for getting back to me, it indeed seems to work from my testing in the Sentry profiling SDK, with some quirks so far (I didn't get stack trace method resolution working yet, but from my understanding it should work because all the info has already come from the runtime). Also the ActivityComputer (and others) don't seem to work when streaming at the moment so that would also need some tuning I guess. I was able to get around them for now but I think they're useful so I'll likely get back to them in the future.

> 1. It would be worthwhile to architect this similiar to how we do with `TraceEventSession` for live sessions.  This allows the `TraceLog` to know that it is being used in a live mode instead of a file-based mode.  There are some differences in how one can use the `TraceLog` when its live, because it hasn't seen the "end" of the trace, unlike in the file-based mode.

Yes, I've made the change by reusing the same code, really, with a minor change of setting up the `kernelSession` in the `TraceEventSession` factory instead of the common constructor.

> 2. Can you share what differences you're seeing?  It's possible that it's just some piece of data that hasn't made it over to the `TraceLog`.

e.g the following line is missing from the captured output but is present in the `eventpipe-dotnetcore2.1-linux-x64-objver3.netperf.baseline.txt`:

```
System.Threading.Tasks.TplEventSource/SetActivityId, 1, <Event MSec=   "111.4517" PID="80749" PName="Process(80749)" TID="80749" EventName="SetActivityId" ProviderName="System.Threading.Tasks.TplEventSource" NewId="10000000-0000-0000-0000-000000000001"/>
```

Also, all the custom events are missing, as present in `eventpipe-dotnetcore2.1-linux-x64-tracelogging.netperf.baseline.txt`

I'm pretty sure it's a test-code issue, because as I've mentioned earlier, the overall approach seems to work in general, based on my changes to the profiling code in the Sentry SDK which now starts the profiler early in the process and slices profiles when actually needed (drops the other samples if not currently needed). I think this functionality will need better tests here, instead of trying to reuse what was already present in `EventPipeParsing/Streaming`.

**Comment by vaind** [comment]
So the issue I'm having with stacktrace info not being available is due to missing modules & method info because the rundown on a session only happens when a session is stopped, while I'm trying to access this info while the sample-profiler is running (on each sample). I've raised an inquiry for this issue: https://github.com/dotnet/runtime/issues/86103

I was wondering whether I could start & stop a session, in the beginning, to force the rundown to happen and then use these events to feed TraceLog. Afterwards, the runtime information would rely on the runtime provider to get updates. @brianrob do you think this could work, conceptually or do you see any issues that would need to be overcome (or prevent this completely)?

**Comment by vaind** [comment]
> I was wondering whether I could start & stop a session, in the beginning, to force the rundown to happen and then use these events to feed TraceLog. Afterwards, the runtime information would rely on the runtime provider to get updates. @brianrob do you think this could work, conceptually or do you see any issues that would need to be overcome (or prevent this completely)?

So I've verified this works fine, I'll just need to add test cases to this repo so that it's covered here.

**Comment by vaind** [comment]
@brianrob Any chance you could have a first look at `TraceLog.cs` changes if they seem reasonable? If so, I'd update tests to cover the changes & mark this PR as Ready for review

**Comment by brianrob** [comment]
@vaind, apologies for not getting to this sooner.  I am reviewing your code.

**Comment by brianrob** [comment]
@vaind, the concept here looks good.  One thing I think we will need to resolve before this is ready for review is the rundown design.  Your instinct to trigger a rundown at the beginning of the trace is the right one.  I was going to recommend that you use the `StartRundown` keyword to do this, but it doesn't look like this is plumbed through the runtime, so it wouldn't have any impact.  If I recall correctly, EventPipe has a relatively hardcoded rundown setup, in which case we may not have many options here.

@davmason, do you know if rundown is configurable at all?  If I recall correctly, end rundown is hardcoded at the end of the session.  Is that right?  I'm trying to figure out the best option for @vaind to trigger a start rundown at the beginning of streaming so that `TraceLog` gets method rundown information and can resolve stacks.

@vaind, once this is resolved, then I think it will make sense to review this more.  I saw some diffs that I want to examine further, but I'm not on a great network connection today and so I don't have access to a quality diff tool.

**Comment by vaind** [comment]
> @vaind, the concept here looks good. One thing I think we will need to resolve before this is ready for review is the rundown design. Your instinct to trigger a rundown at the beginning of the trace is the right one. I was going to recommend that you use the `StartRundown` keyword to do this, but it doesn't look like this is plumbed through the runtime, so it wouldn't have any impact. If I recall correctly, EventPipe has a relatively hardcoded rundown setup, in which case we may not have many options here.

Thanks for getting back to me @brianrob. And yes,  there doesn't seem a way to trigger a rundown on the same session. I've discussed that in the runtime issue, I've created two weeks ago: https://github.com/dotnet/runtime/issues/86103

**Comment by davmason** [comment]
> @davmason, do you know if rundown is configurable at all? If I recall correctly, end rundown is hardcoded at the end of the session. Is that right? I'm trying to figure out the best option for @vaind to trigger a start rundown at the beginning of streaming so that `TraceLog` gets method rundown information and can resolve stacks.

Rundown is not currently configurable with EventPipe. You are remembering correctly, it is hardcoded to run at the end of the session. The internal code has the option to configure whether to do rundown or not, but we don't expose it over the IPC commands so customers can't turn it off or on. It would be fairly easy to expose the option to have rundown or not, and I don't think it would be that much work to have it at the start of a trace. I think we could reuse all the existing code

**Comment by vaind** [comment]
> @vaind, once you're ready, go ahead and mark this as ready for review and I'll take a more thorough pass on it.

I've done so. It does need tests but I want to make sure the approach and APIs are OK-ish before trying to devise tests for them.

**Comment by vaind** [comment]
P.S. I've tested it as a consumer in [Sentry .NET SDK](https://github.com/getsentry/sentry-dotnet/pull/2385). What I meant by needing tests is that it needs tests in this repo.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:227]: I think that this function should be private.  There is already a path to create a `TraceLog` from a file path, and I would like to make sure that we ensure that someone doesn't pass an `EventPipeEventSource` from a file into this method.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:256]: @davmason, will this work to ensure that the entire end rundown is received?

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:763]: I think we will want to do some real-time ETW testing here to ensure that nothing has regressed in that path.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:1289]: It would be good to have a comment here that describes why this is necessary.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:10834]: It looks like there are a couple of these asserts that were changed to remove `log.IsRealTime`.  Can you tell me more about what changed here?

**Review by brianrob — COMMENTED** [review]
Thanks for your patience.  Here's a first cut at the review.  I think we may have a few iterations, but overall, this is looking good.

File: src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs
Comment (vaind) [reply src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs:3]: It was unused in the end. Removed.

File: src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs
Comment (vaind) [reply src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs:8]: It was unused in the end. Removed.

**Review by vaind — COMMENTED** [review]
I've made some changes as requested and replied to some questions. 

Keeping the testing open until the questions are resolved.

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:735]: There doesn't seem to be any test case for `CreateFromTraceEventSession()`, i.e. the ETW one - how do you think it would be best to go around testing this (and the EventPipeSession factory too)?

File: src/TraceEvent/TraceLog.cs
Comment (davmason) [reply src/TraceEvent/TraceLog.cs:234]: Yeah, just stopping the session is enough

File: src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs
Comment (rapetum228) [inline src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs:44]: 1000 - magic number

**Comment by vaind** [comment]
> I've made some changes as requested and replied to some questions.
> 
> Keeping the testing open until the questions are resolved.

@brianrob Have you been able to finish reviewing this? Any more concerns or do you think it would be acceptable after adding tests for the changes/new APIs? Also, do you have any preferences for the testing approach?

File: src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs
Comment (bruno-garcia) [inline src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs:44]:
```suggestion
                        FirstSerializedSample = new String(data.ToString().Replace("\n", "\\n").Replace("\r", "\\r").Take(1000).ToArray())
```

Comment (vaind) [reply src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs:44]: Old code, moved from here: https://github.com/microsoft/perfview/pull/1867/files/eb41100b9dfd53913bea8a11a302630d4b407afb#diff-c94b1f053e78f716a557a5876766fc48558133ec18fe0fa752e5cf40c23a6fcfL61

File: src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs
Comment (bruno-garcia) [inline src/TraceEvent/TraceEvent.Tests/Parsing/EventPipeParsing.cs:24]: Oh the type was actually already in

**Comment by bruno-garcia** [comment]
@brianrob this would really help unblock some work we want to do for .NET here at Sentry. Another round of review much appreciated

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:258]: This makes sense because rundown is limited to what's called "stop rundown" - rundown that is triggered when we stop tracing.  You can safely remove this commented out code.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:285]: It should be fine to use the JITTED method calls for all MethodDCStopVerbose events.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [reply src/TraceEvent/TraceLog.cs:735]: Good question - there aren't any tests for this.  We don't really have an environment where we can run such tests elevated in an automated fashion.

I suspect the best thing to do here is to write some tests that emit events while a real-time session of each type is live, and then confirm that those events come through.

The EventPipe one can be automated if we so desire, because it doesn't require admin privileges.  The ETW one will likely just need to remain adhoc.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:4473]: Can you describe the reason for this?  Just want to make sure I understand if this is due to a bug or design change.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:9244]: Same here - is this a bug?  I could believe that it is.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [reply src/TraceEvent/TraceLog.cs:10808]: Gotcha.  Go ahead and remove it then.  If we notice this isn't the case, we can always re-add it.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [reply src/TraceEvent/TraceLog.cs:205]: Yes, you're right - you can't re-use the session as far as I can tell.  Barring some change in EventPIpeSession or the DiagnosticsClient, I think we'll have to do something here that eventually changes.

It occurs to me that the rundown session needs to be configured properly.  I wonder if we should have another construct in TraceEvent that handles this.  We could do something like this:

Create an EventPipeRundownConfiguration class that has two static factories:
 - None (creates a configuration that does not trigger rundown)
 - Enable (takes a DiagnosticsClient and creates a rundown session)

This would allow us to control how rundown occurs later on, and would allow us to make changes to use StartRundown without users having to change their code.  It's not exactly what I was hoping for, but maybe OK.  What do you think?

**Comment by brianrob** [comment]
> > I've made some changes as requested and replied to some questions.
> > Keeping the testing open until the questions are resolved.
> 
> @brianrob Have you been able to finish reviewing this? Any more concerns or do you think it would be acceptable after adding tests for the changes/new APIs? Also, do you have any preferences for the testing approach?

Just did a pass.  This is looking good.  I added a bit of comment on testing, but not sure if that's what you're looking for.  Let me know and I can think on this some more.  Thanks.

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:4473]: I'd say it was a bug here - after Process() exited, you wouldn't have received all the events yet because the realtime queue has a flush interval.

I've noticed this issue at a certain time when the EventPipe was also routed through this queue. On the other hand, it doesn't use it anymore so the fix here isn't really helping **me**, but it still is a fix, although unrelated to the changes in this PR...

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:285]: OK, changed the TODO to a comment and cleaned up  there.

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:258]: thx, removed

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:205]: Sounds good, I've made the change

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:9244]: I can't confirm this is a bug in general, but for the realtime processing that happens with EventPipe session, it is necessary to let the process be created.

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:735]: Alright, I'll implement the EventPipe session tests similar to what I've done in Sentry dotnet SDK - there should be no problem running them in CI.

Comment (vaind) [reply src/TraceEvent/TraceLog.cs:727]: I've added the test. After much headscratching why it was always freezing without an error I've found out that's what happens when you try starting an EventPipe session on a DiagnosticClient connection to a process that is older than .NET core 3.0... 
Since I was trying to connect to the test process itself to test the session, I had to add another `targetframework` to the test project and run the new test conditionally only on the newer .NET. I hope that's alright.

Comment (brianrob) [reply src/TraceEvent/TraceLog.cs:727]: Yes, that's great.  I don't think the IPC channel exists on .NET Core 3.x, so I wouldn't expect it to work there.

**Comment by brianrob** [comment]
FYI, I am in the middle of reviewing this, but need to stop right now.  Things are looking good though - I need to look at the test changes and the CI failures.

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:727]: It does work since .NET core 3.0. It's only missing on .NET framework.

Also, since I've found out the .netcoreapp3.1 is already being installed in the existing azure pipelines CI workflow, I've adapted the targetframework from net6.0 to netcoreapp3.1 (already in use in `PerfViewCollect` project

**Comment by vaind** [comment]
> FYI, I am in the middle of reviewing this, but need to stop right now. Things are looking good though - I need to look at the test changes and the CI failures.

I've tried a couple of things to get this to work but am always hitting a roadblock.  
Compilation is fixed by switching to .NET 6.0 in the UseDotNet task - it still seems to compile for net462 just fine. Not sure what the task does exactly as I don't have experience with Azure pipelines.
However, tests then start to fail (test runner crashes). I've figured out, this is caused by WPF tests in `StackWindowTests.cs`. Since I can't find a clean solution without drawbacks, I think it will need some input from you which way you'd want to go from here.

**Comment by brianrob** [comment]
FYI I am doing some experiments to see if I can help unblock the tests: https://github.com/microsoft/perfview/pull/1894.

**Comment by brianrob** [comment]
In my testing over in #1894 I discovered that running each of the test assemblies separately works.  It seems that the combination of all of the test runs is an issue.  I'm wondering if the mix of .NET Framework and .NET Core is the issue - I don't have a clear picture of exactly how that's happening on the test side.  I wonder if we need to find a way to do an explicit .NET Core test run and an explicit .NET Framework test run, and choose the specific assemblies that should run for each.  This would allow us to avoid targeting the WPF tests at .NET Core, and would allow us to make sure that we're not attempting to run .NET Framework code against .NET Core and vice versa.  Thoughts?

**Comment by vaind** [comment]
> In my testing over in #1894 I discovered that running each of the test assemblies separately works. It seems that the combination of all of the test runs is an issue. I'm wondering if the mix of .NET Framework and .NET Core is the issue - I don't have a clear picture of exactly how that's happening on the test side. I wonder if we need to find a way to do an explicit .NET Core test run and an explicit .NET Framework test run, and choose the specific assemblies that should run for each. This would allow us to avoid targeting the WPF tests at .NET Core, and would allow us to make sure that we're not attempting to run .NET Framework code against .NET Core and vice versa. Thoughts?

In that case, how about merging this PR first and then there's not so much duplication going on: https://github.com/microsoft/perfview/pull/1896

**Comment by bruno-garcia** [comment]
@brianrob @vaind wondering if there's any update on this? Thanks

**Comment by ericsampson** [comment]
bump?

**Comment by vaind** [comment]
Hi @brianrob, I've updated this PR with the latest changes from the main branch and all the CI passes. Is there anything I should do to so that these changes can eventually land?

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:227]: Please add comments to the public APIs here.

File: src/TraceEvent/TraceLog.cs
Comment (brianrob) [inline src/TraceEvent/TraceLog.cs:322]: This should not happen because we'll always be attached to a real process, and there is no idle process for event pipe.  You can remove this check.

**Review by brianrob — COMMENTED** [review]
@vaind thank you for your patience on this and for all of the work that you've done thus far.  I think we're getting close.  I've added a few comments here, but I don't think any are big.

I see the following as the path to merge:

1. @noahfalk, can I get you to give this a review please?  This is for livesession support for EventPipe.
2. I would like to validate that we don't regress live session support for ETW traces.  Have you done any testing in this space?  I know we don't have tests here in the repo, but presumably, we can do some adhoc testing, similar to https://github.com/brianrob/examples/tree/main/src/realtime-session-with-stacks.

Assuming we can resolve this, I think we'll be ready to merge (or at least know what we need to do).

**Review by noahfalk — APPROVED** [review]
This looked fine to me, but I warn my knowledge of the TraceLog portion of TraceEvent is limited so I wouldn't read too much into it. If there were anything subtly wrong I don't expect I would catch it without spending a while researching the internals of TraceLog.

**Comment by bruno-garcia** [comment]
wooo thanks @noahfalk !

Getting there

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:322]: right, thanks for the insight

File: src/TraceEvent/TraceLog.cs
Comment (vaind) [reply src/TraceEvent/TraceLog.cs:230]: docs added to the class & its public methods

**Comment by vaind** [comment]
@brianrob thanks for the review, somehow I've missed it until now.

I've addressed your requests and updated the branch. Let me know if there's anything else needed.

**Comment by vaind** [comment]
P. S. besides review changes, there was one more fix [`36d2e2c` (#1867)](https://github.com/microsoft/perfview/pull/1867/commits/36d2e2c28c1ebeaa780908dc6f831144872a37a5#diff-31ffbfdad96fc84f3aa7d25e1980cdca5a4d7870529c669ac9cc975cf24969c8R342) - the structures would grow indefinitely otherwise (https://github.com/getsentry/sentry-dotnet/issues/3375)

File: src/PerfView.Tests/StackViewer/StackWindowTests.cs
Comment (vaind) [reply src/PerfView.Tests/StackViewer/StackWindowTests.cs]: Sorry about this, must have been some temp change when testing this and it slipped through because the file showed up as collapsed in the GH PR. reverted to the original

**Comment by bruno-garcia** [comment]
This PR had its 1 year anniversary 😅 

It's got all the review points addressed. Could we please get some help pushing this through 🙏

**Comment by ericsampson** [comment]
@bruno-garcia too bad that Elon tanked Twitter 😝

File: src/TraceEvent/TraceEvent.csproj
Comment (brianrob) [inline src/TraceEvent/TraceEvent.csproj:54]: This will also require a corresponding addition to `<NuspecProperties>` above, as well as an entry in the dependencies section of the nuspec file.  This is what ensures that NuGet restores the dependency.

**Review by brianrob — COMMENTED** [review]
Looking good.  One oustanding issue and then I think we're about ready to merge.

**Comment by brianrob** [comment]
> This PR had its 1 year anniversary 😅
> 
> It's got all the review points addressed. Could we please get some help pushing this through 🙏

Yup!  Happy Anniversary?

File: src/TraceEvent/TraceEvent.csproj
Comment (vaind) [reply src/TraceEvent/TraceEvent.csproj]: missed this. it's done now

**Comment by vaind** [comment]
FYI I'm investigating an unbound memory growth when streaming EventPipe. At first, I suspected the cause is this [`GC.KeepAlive`](https://github.com/getsentry/perfview/blob/18df88f900a562f38f6df1500c7ac0ab0a58be6c/src/TraceEvent/EventPipe/EventCache.cs#L198) but it seems there's (also) another issue. Maybe it has to do with pinned buffers in or in EventCache or something else entirely. In any case, I'm unable to pinpoint the issue, mainly because in while memory profiling, the managed memory actually doesn't grow. Or more specifically, managed memory grows, but when a GC is triggered, it is freed but the unmanaged memory stays the same. 
image

On the other hand, if I run the same app with ` procgov64 --maxmem 70M -- app-name.exe` (https://github.com/lowleveldesign/process-governor) the memory usage is stable (with an occasional crash which may be caused by some temporary memory pressure I guess). This, to me, sounds like it may have something to do with how runtime reuses (or doesnt) heaps when there are pinned objects inside... Any ideas? In any case I wouldn't want to block this PR so feel free to merge it if you're fine with the current status and any changes can come in  a followup.

**Comment by brianrob** [comment]
I don't expect the `GC.KeepAlive` call to be a problem.  That's just ensuring that the JIT reports the object as live until the call to `GC.KeepAlive` completes.

It's hard to say where the unmanaged memory growth is coming from.  Does it repro in a minimal repro app?  It might be worth looking at a GC trace (GCStats in PerfView) to see if the GC heap is increasing in size and this is just fragmentation being held, or if there is truly a native leak.

Either way, I am OK taking this change as is, and we can address this in a follow-up.

**Review by brianrob — APPROVED** [review]
Thanks for your patience and for bringing this across the finish line!

**Pull request merged**
